theano/sandbox/gpuarray/basic_ops.py

```diff
@@ -259,57 +329,59 @@ def c_code_cache_version(self):
 
     def grad(self, inputs, grads):
         gz, = grads
-        return [gpu_from_host(gz)]
+        return [GpuFromHost(inputs[0].type.context_name)(gz)]
 
     def R_op(self, inputs, eval_points):
         ev, = eval_points
-        if isinstance(ev, tensor.TensorType):
-            return [gpu_from_host(ev)]
-        else:
-            return [ev]
+        return self(ev)
```

Comment:
@lamblin conviced me that this is the right way to implement R_op and that the previous way was bad and only worked ok since the jacobian is effectively the identity and we are patching up for things that are not in the expected location everywhere.

I can do a similar change in the old backend as well.